core/src/Cabana_BufferedAoSoA.hpp

```diff
@@ -0,0 +1,270 @@
+// allocated without running OOM 4) User can provide a hint for ideal number of
+// buffers to chunk the data into while buffering
+template <int max_buffered_tuples,
+          int requested_buffer_count, // TODO: should this be an optional hint,
+                                      // and not be prescriptive
+          class Exec_Space,           // TODO: we can infer this?
+          class AoSoA_t>
+class BufferedAoSoA
+{
+    // TODO: make things private
+  public:
+    using AoSoA_type = AoSoA_t;
+
+    using slice_tuple_t = UnpackTuple<AoSoA_t, typename AoSoA_t::member_types>;
+    slice_tuple_t slice_tuple;
+
+    // TODO: this has nothing to do with the class, only the
+    // specialized get_slice does
+    /**
+     * @brief Underlying generic style getter to obtain a reference to
+     * i-th item in the slice tuple
+     *
+     * @param tuple The tuple to access from
+     *
+     * @return The i-th value in the tuple
+     */
+    template <typename _AoSoA_t, std::size_t i, typename HeadItem,
+              typename... TailItems>
+    KOKKOS_INLINE_FUNCTION typename TupleLeaf<_AoSoA_t, i, HeadItem>::value_t &
+    _Get( TupleImpl<_AoSoA_t, i, HeadItem, TailItems...> &tuple )
+    {
+        return tuple.TupleLeaf<AoSoA_t, i, HeadItem>::value;
+    }
+
+    /**
+     * @brief Getter to access the slices this class generates for the
+     * user
+     *
+     * @return Valid slice for the current buffer at position i
+     */
+    template <std::size_t i>
+    KOKKOS_INLINE_FUNCTION typename AoSoA_t::template member_slice_type<i> &
+    get_slice()
+    {
+        return _Get<AoSoA_t, i>( slice_tuple );
+    }
+
+    // TODO: should this be a more heap based / dynamic allocation?
+    // TODO: should we place these in the "target" memory space, not
+    // the default?
+    AoSoA_t internal_buffers[requested_buffer_count];
+
+    BufferedAoSoA( AoSoA_t &original_view_in )
+        : slice_tuple( internal_buffers[0] )
+        , original_view( original_view_in )
+        , buffer_size( max_buffered_tuples )
+    {
+        // TODO: this is only used internally now, and can likely be a
+        // non-pointers
+        last_filled_buffer = -1;
+
+        // TODO: We may we want to override the user on their requested
+        // values if they don't make sense? Both num_buffers and buffer_size
+        num_buffers = requested_buffer_count;
+
+        std::cout << "The size of the passed view is "
+                  << original_view_in.size() << std::endl;
+        std::cout << "The size of our view is " << original_view.size()
+                  << std::endl;
+
+        // Resize the buffers so we know they can hold enough
+        for ( int i = 0; i < num_buffers; i++ )
+        {
+            // internal_buffers[i].resize(buffer_size);
+            std::cout << "Making buffer of size " << buffer_size << std::endl;
+            std::string internal_buff_name = "internal_buff " + i;
```

Comment:
this does not do what you want

Reply:
Ha, you're right. Good catch. Too much python on the weekends..